### ISSUE TITLE
[C10D] Make collectives backwards throw an error

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#152127]
Today functional collectives and C10D collectives silently ignore backwards, which can surprise users and lead to missing gradients and incorrect training.

Many users of these collectives do not intend to use the backwards pass, so this limitation is not affecting them.  They either call functional_collectives _during_ the backward pass, in an explicit no_grad context, (e.g. DDP, FSDP) or they write a custom autograd.Function that performs some explicitly chosen collectives as part of forward pass and backward pass.

Other users may use the collective directly during forward, and expect to have proper autograd support.  These users would observe silent correctness problems.

We should explicitly register a backwards kernel for functional collectives that throws an error.  Later, we can replace this kernel with an actual backwards implementation, but we don't have to do that all at once and there are some more other design decisions to make regarding performance of the backwards pass, so we should plug this hole first.

**Comment by wconstab** [comment]
Talking to @bdhirsh about this, it sounds like we'd have to write an autograd.function for each collective in order to implement its backward error.  I was hoping for a way to register one 'backwards error kernel' to all the collectives quickly, but it doesn't appear to be easy to do that.  So we may still do this but if its just as much work to implement the backwards formulas we can do that instead.
